File: src/main/java/service/GenerationService.java
package service;

import java.util.HashSet;
import java.util.Set;
import lombok.AllArgsConstructor;
import model.Matrix;
import model.MatrixPath;

@AllArgsConstructor
public class GenerationService {

    private static final int TRESHOLD = 1_000_000;

    private final MatrixPathGenerator matrixPathGenerator;
    private final MatrixPathSimplifier matrixPathSimplifier;

    public Set<MatrixPath> newGeneration(Matrix matrix, int size) {
        Set<MatrixPath> res = new HashSet<>();

        int failureCounter = 0;
        while (res.size() < size) {
            var matrixPath = matrixPathGenerator.generatePath(matrix);
            MatrixPath simplifiedPath = matrixPathSimplifier.simplify(matrix, matrixPath);
            boolean isNewValue = res.add(simplifiedPath);
            if (!isNewValue) {
                ++failureCounter;
            }
            if (failureCounter > TRESHOLD) {
                System.out.println("Matrix is too pidorskaya, can't build required amount of unique path");
                return res;
            }
        }

        return res;
    }
}

File: src/main/java/service/GenerationService.kt
package service

import model.Matrix
import model.MatrixPath
import java.util.HashSet

class GenerationService(
    private val matrixPathGenerator: MatrixPathGenerator,
    private val matrixPathSimplifier: MatrixPathSimplifier
) {

    fun newGeneration(matrix: Matrix?, size: Int): Set<MatrixPath> {
        val res: MutableSet<MatrixPath> = HashSet()
        var failureCounter = 0
        while (res.size < size) {
            val matrixPath = matrixPathGenerator.generatePath(matrix)
            val simplifiedPath = matrixPathSimplifier.simplify(matrix, matrixPath)
            val isNewValue = res.add(matrixPath)
            if (!isNewValue) {
                ++failureCounter
            }
            if (failureCounter > TRESHOLD) {
                println("Matrix is too pidorskaya, can't build required amount of unique path")
                return res
            }
        }
        return res
    }

    companion object {
        private const val TRESHOLD = 1000000
    }
}
